src/app/cli/src/init/test_submit_to_archive.ml

```diff
@@ -419,30 +419,89 @@ let mk_payment ~(valid_until : Mina_numbers.Global_slot_since_genesis.t)
-    ~constraint_constants keypair : User_command.Valid.t Sequence.t list =
+(** Create a transaction to create [n_accounts] many zkapp accounts and insert them into 
+    the accounts table.
+*)
+let create_zkapp_accounts_tx ?(fee = 20_000_000) ?(initial_balance = 10)
+    ~constraint_constants ~sender ~nonce_ref ~n_accounts ~account_state_tbl =
+  let zkapp_keypairs =
+    List.init n_accounts ~f:(fun _ -> Signature_lib.Keypair.create ())
+  in
+  let (zkapp_command_spec : Transaction_snark.For_tests.Deploy_snapp_spec.t) =
+    { sender = (sender, !nonce_ref)
+    ; fee = Currency.Fee.of_nanomina_int_exn fee
+    ; fee_payer = None
+    ; amount = Currency.Amount.of_mina_int_exn initial_balance
+    ; zkapp_account_keypairs = zkapp_keypairs
+    ; memo = Signed_command_memo.create_from_string_exn "Zkapp create accounts"
+    ; new_zkapp_account = true
+    ; snapp_update = Account_update.Update.dummy
+    ; preconditions = None
+    ; authorization_kind = Signature
+    }
+  in
+  let%map tx =
+    Transaction_snark.For_tests.deploy_snapp ~constraint_constants
+      ~signature_kind:Testnet zkapp_command_spec
+  in
+  let () = nonce_ref := Unsigned.UInt32.succ !nonce_ref in
+  let () =
+    List.iter zkapp_keypairs ~f:(fun zkapp_keypair ->
+        let zkapp_pk =
+          Signature_lib.Public_key.compress zkapp_keypair.public_key
+        in
+        let zkapp_account_id = Account_id.create zkapp_pk Token_id.default in
+        let zkapp_account_base =
+          Account.create zkapp_account_id
+            (Currency.Balance.of_mina_int_exn initial_balance)
+        in
+        let zkapp_account =
+          { zkapp_account_base with zkapp = Some Zkapp_account.default }
+        in
+        Account_id.Table.set account_state_tbl ~key:zkapp_account_id
+          ~data:(zkapp_account, `Ordinary_participant) )
+  in
+  tx
+
+let generate_txs ~valid_until ~nonce_ref ~n_zkapp_txs ~n_payments ~n_blocks
+    ~constraint_constants ~max_cost ~account_state_tbl ~vk ~genesis_constants
+    keypair : User_command.Valid.t Sequence.t list =
   let signer_pk = Public_key.compress keypair.Keypair.public_key in
   let event_elements = 12 in
   let action_elements = 12 in
-  let generate_payments block_i =
+  let generate_payments () =
     Sequence.init (n_payments + n_zkapp_txs) ~f:(fun i ->
-        let nonce =
-          Mina_numbers.Account_nonce.(
-            add init_nonce @@ of_int
-            @@ ((block_i * (n_payments + n_zkapp_txs)) + i))
-        in
         let command =
           if i < n_payments then
             (* Creates a simple payment that initializes a new account *)
-            User_command.Signed_command
-              (mk_payment ~valid_until ~nonce ~signer_pk keypair)
+            let res =
+              User_command.Signed_command
+                (mk_payment ~valid_until ~nonce:!nonce_ref ~signer_pk keypair)
+            in
+            let () = nonce_ref := Unsigned.UInt32.succ !nonce_ref in
```

Comment:
Nit: `Ref.replace nonce_ref Unsigned.UInt32.succ;`

Reply:
Also, all branches bumped the nonce. it make sense to lift nonce bumping out of the big if.

Reply:
Agreed, but I'm just going to keep them, hopefully as future idiot-proofing